src/Platforms/DeviceManager.h

```diff
@@ -0,0 +1,89 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers.
+//
+// File developed by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
+//
+// File created by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
+//
+//////////////////////////////////////////////////////////////////////////////////////
+
+
+#ifndef QMCPLUSPLUS_DEVICEMANAGER_H
+#define QMCPLUSPLUS_DEVICEMANAGER_H
+
+#include <memory>
+#include <config.h>
+#if defined(ENABLE_CUDA)
+#include "CUDA/CUDADeviceManager.h"
+#endif
+#if defined(ENABLE_OFFLOAD)
+#include "OMPTarget/OMPDeviceManager.h"
+#endif
+
+namespace qmcplusplus
+{
+
+/** orchestrate platforms (programming models) and set up the default device on each platform
+ *
+ * Currently support CUDA, OpenMP
+ *
+ * Each platform provides its own device manager class. Each individual device manager
+ * initializes all the devices visible to the process and also set the default device.
+ * The first initialized platform selects the default device for the current process and all the rest
+ * platforms follow that choice.
+ *
+ * The device manager on each platform may be constomized to store a list of known device numbers or
+ * a list of context device handles depending on the need.
+ *
+ * DeviceManager assumes there is only one type of accelerators although they may support multiple
+ * platforms. Under this assumption, the numbers of devices captured on all the platforms must agree.
+ *
+ * DeviceManager::global is intended to the per-process singleton and should be initialized after
+ * MPI initialization in main().
+ */
+class DeviceManager
+{
+public:
+  /* constructor
+   * @param local_rank the rank id of the node local MPI communicator
+   * @param local_size the size of the node local MPI communicator
+   */
+  DeviceManager(int local_rank, int local_size);
```

Comment:
I didn't completely get how this class worked when we talked yesterday, I assumed there was limited access to the constructor. In the example I gave yesterday the get() function goes into the singleton class and you make the constructor private. The reason you put the one static copy of your class in a local function is that "Initialization of function-local statics is guaranteed to occur only once even when called from multiple threads, and may be more efficient than the equivalent code using std::call_once."

The pattern is attributed to Scott Meyers. The singleton is constructed on first access. Since singletons are by convention only accessed through their acessor or the reference it returns there is no "trouble." There can be trouble in certain linkage scenarios the one I know if is if the calls to the singleton are in header files which are included from different compilation units.

Reply:
std::call_once seems solving a different problem I'm looking at.
Now I changed the static DeviceManager::global to private.
initializeGlobalDeviceManager is for initialization
If being called multiple times, the second and later calls will throw error.
getGlobal() is for access the global instance.
It checks if global is nullptr.

Reply:
I'm satisfied with this. It's not threadsafe but it is good enough for us. The general rule in QMCPACK is that thread safety is the caller's responsibility. I will catch bad single thread application or test program logic which I think is the most likely issue.

There is nothing wrong with std::call_once (although I doubt it is compatible with openmp) for situations where you have a single resource that you are going to initialize lazily and you know that will be from a potentially concurrent context it works fine. You could easily build an abstraction that either used the stdlib object or implemented the functionality in openmp. I'm sure there are other uses as well.